lib/synapse/config_generator/haproxy.rb

```diff
@@ -1055,9 +1072,28 @@ def generate_backend_stanza(watcher, config)
             @restart_required = true
           end
         end
+
         backends[backend_name] = backend.merge('enabled' => true)
+
+        # If the the registry defines the haproxy_server_id that must be preferred.
+        server_opts = (backend['haproxy_server_options'] || 'no match').split(' ')
+        @server_id_map[watcher.name][backend_name] = server_opts[server_opts.index('id') + 1].to_i if server_opts.include?('id')
+        # This is going to be an issue ... and haproxy will fail if this happens ... oh well
+        @server_id_map[watcher.name][backend_name] = backend['haproxy_server_id'] if backend['haproxy_server_id']
       end
 
+      # Now that we know the maximum possible existing haproxy_server_id for
+      # this backend, we can set any that don't exist yet.
+      watcher.backends.each do |backend|
+        backend_name = construct_name(backend)
+        @server_id_map[watcher.name][backend_name] ||= (
+          ((@server_id_map[watcher.name].values + [0]).compact.max) % @max_server_id + 1
```

Comment:
Hm, I'm less concerned about someone setting manual ids and more concerned with holes created by a churning server (e.g. srv1 has id 1, and then you have a container start and stop 65k times until it loops over), but I see what you mean.

I do wish that ruby had a better stdlib (heap or sortedset without rbtree), but tbh I don't think that having an O(n) probe here is _that_ bad. N is guaranteed to be reasonably small (65k) and we only end up doing it when we have a new server. Sorted set does seem like exactly what I want ... but ... it's slow ... I'll figure something out.